### PR TITLE
[JavaScript] new concept exercise - `callbacks`

### PR DESCRIPTION
**Pull request by neenjaw**
This is an early draft, I'm looking for for some feedback regarding the scope of this exercise as I am working through the after document.

1) This exercise will be run in the node environment, can I make use of some sort of node API so that I have have the student handle an async response?  If not, thoughts or suggestions to mock?

2) Can I reference certain types of callbacks:
   - event listeners attached to DOM objects in the browser context
   - system api's for the node context

### CODE REVIEW

**Comment by ErikSchierboom** [comment]
@neenjaw I'd prefer @SleeplessByte to look at this. He's unavailable at the moment, so this might take a little time before it gets reviewed.

**Comment by neenjaw** [comment]
@SleeplessByte, thoughts?

**Comment by SleeplessByte** [comment]
@neenjaw still need me on this? we talked a bit "offline", but we can pick it up again

**Comment by neenjaw** [comment]
I'm working on it slowly, keep getting distracted.

File: languages/javascript/exercises/concept/fruit-picker/fruit-picker.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { checkStatus, checkInventory } from './grocer';
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/fruit-picker.js:1]:
```suggestion
// @ts-check

import { checkStatus, checkInventory } from './grocer';
```

**Review by SleeplessByte — COMMENTED** [review]
In general this is close to be done.

See inline commentary.

asynchronous callbacks are missing. Follow up exercise?

File: languages/javascript/exercises/concept/fruit-picker/.docs/instructions.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+The grocer's API provides a function to query their inventory called `checkInventory`. It receives two arguments: a _query_, and a _callback_ function.
+
+The query takes the form of an _object_:
+
+```javascript
+const query = {
+  variety: <string>,
+  quantity: <number>
+}
+```
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply languages/javascript/exercises/concept/fruit-picker/.docs/instructions.md:26]: Oh, I didn't even think to use typescript codeblock hinting for that!

**Comment by neenjaw** [comment]
Yes, I didn't have a good way to test asynchronous callbacks, it could be a follow-up exercise

File: languages/javascript/exercises/concept/fruit-picker/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+applyToSquare(function squarePerimeter(side) {
+  return side * 4
+})
+```
+Or an anonymous inline arrow function expression:
+
+```javascript
+applyToSquare((side) => side * 4)
+```
+// The argument "(side) => side * 4" is the callback
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/.docs/introduction.md:33]:
````suggestion
// The argument "(side) => side * 4" is the callback
```
````

File: languages/javascript/exercises/concept/fruit-picker/grocer.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/**
+ * STORE STATUS API
+ */
+
+let STATUS = 'OFFLINE';
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/grocer.js:5]: Should not be a constant (STATUS signals constant).

File: languages/javascript/exercises/concept/fruit-picker/grocer.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ */
+export function checkStatus(callback) {
+  return callback(STATUS);
+}
+
+/**
+ * INVENTORY API
+ */
+
+let LAST_QUERY = undefined;
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/grocer.js:34]: Same

File: languages/javascript/exercises/concept/fruit-picker/grocer.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+let RESPONSE = undefined;
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/grocer.js:35]: Same

**Review by SleeplessByte — APPROVED** [review]
Few nitpicks. I like it a lot.

**Comment by SleeplessByte** [comment]
> Yes, I didn't have a good way to test asynchronous callbacks, it could be a follow-up exercise

We can do it together.

File: languages/javascript/exercises/concept/fruit-picker/fruit-picker.spec.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
+import {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/fruit-picker.spec.js:1]:
```suggestion
// @ts-check

import {
```

See if vscode still likes this file if you add this

File: languages/javascript/exercises/concept/fruit-picker/fruit-picker.spec.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  getLastQuery,
+  resetQuery,
+} from './grocer';
+
+describe('service status', () => {
+  beforeEach(() => {
+    resetStatus();
+  });
+
+  test('returns the reset status of the service', () => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/fruit-picker/fruit-picker.spec.js:20]:
```suggestion
  test('returns the initial status of the service', () => {
```

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/fruit-picker/fruit-picker.spec.js:20]: I know it's technically reset status, but resetting is not "seen" by the user

**Review by SleeplessByte — APPROVED** [review]

**Pull request merged**